File: examples/smart-router/policy_server.go
// Smart-router policy sidecar for busbar (`route: webhook`), in Go.
//
// Busbar POSTs a projection of the request + the pool's candidates before each
// request's failover loop; this sidecar classifies the request into a task
// bucket and returns a ranked preference {"order":[idx,...]}. It uses ONLY the
// fields busbar actually sends (see src/routing/webhook.rs): no prompt text ever
// leaves busbar, so classification is on shape (size, counts, tools, streaming,
// max_tokens), not content.
//
// Fail-safe: if this process is slow, down, or wrong, busbar coerces the decision
// to the pool's on_error (default: weighted) after policy.timeout_ms (default
// 1ms by default; raise it for hooks that do I/O). A broken sidecar never blocks a request.
//
// Build/run:  go run policy_server.go [addr]   (default 127.0.0.1:8787)
// Stdlib only, no dependencies.
package main

import (
	"encoding/json"
	"log"
	"net/http"
	"os"
	"sort"
)

// What busbar sends. Absent numeric signals arrive as JSON null → nil pointer.
type request struct {
	Pool            string  `json:"pool"`
	IngressProtocol string  `json:"ingress_protocol"`
	MessageCount    int     `json:"message_count"`
	HasTools        bool    `json:"has_tools"`
	TotalChars      int     `json:"total_chars"`
	MaxTokens       *int    `json:"max_tokens"`
	Stream          bool    `json:"stream"`
}

type candidate struct {
	Idx                  int      `json:"idx"`
	Model                string   `json:"model"`
	Tier                 *string  `json:"tier"`
	CostPerMtok          *float64 `json:"cost_per_mtok"`
	LatencyMs            *float64 `json:"latency_ms"`
	AvailableConcurrency int      `json:"available_concurrency"`
	BudgetRemaining      *int64   `json:"budget_remaining"`
	RateHeadroom         *float64 `json:"rate_headroom"`
}

type payload struct {
	Request    request     `json:"request"`
	Candidates []candidate `json:"candidates"`
}

type reply struct {
	Order   []int `json:"order,omitempty"`
	Abstain bool  `json:"abstain,omitempty"`
}

// Per-bucket weights [cost, latency, concurrency] + the tiers to boost.
type weights struct {
	cost, lat, conc float64
	tiers           []string
}

const tierBoost = 0.5

func classify(r request) weights {
	switch {
	case r.HasTools: // agent / code traffic: send it to the frontier model
		return weights{0.20, 0.40, 0.40, []string{"fable"}}
	case (r.MaxTokens != nil && *r.MaxTokens >= 4096) || r.TotalChars > 24000: // long-form: opus territory
		return weights{0.40, 0.20, 0.40, []string{"opus"}}
	case !r.Stream && r.MessageCount <= 1: // single-shot batch: cheapest wins
		return weights{0.60, 0.10, 0.30, []string{"haiku"}}
	default: // a human waiting on an interactive answer: the everyday driver
		return weights{0.30, 0.50, 0.20, []string{"sonnet"}}
	}
}

func contains(xs []string, s *string) bool {
	if s == nil {
		return false
	}
	for _, x := range xs {
		if x == *s {
			return true
		}
	}
	return false
}

func rank(p payload) reply {
	if len(p.Candidates) == 0 {
		return reply{Abstain: true}
	}
	w := classify(p.Request)
	var maxCost, maxLat float64
	var maxConc int
	for _, c := range p.Candidates {
		if c.CostPerMtok != nil && *c.CostPerMtok > maxCost {
			maxCost = *c.CostPerMtok
		}
		if c.LatencyMs != nil && *c.LatencyMs > maxLat {
			maxLat = *c.LatencyMs
		}
		if c.AvailableConcurrency > maxConc {
			maxConc = c.AvailableConcurrency
		}
	}
	// Missing per-candidate signals score neutral (0.5): a cold lane is not punished.
	norm := func(v *float64, max float64) float64 {
		if v == nil || max <= 0 {
			return 0.5
		}
		return 1.0 - *v/max // lower cost/latency is better
	}
	score := func(c candidate) float64 {
		concS := 0.5
		if maxConc > 0 {
			concS = float64(c.AvailableConcurrency) / float64(maxConc)
		}
		s := w.cost*norm(c.CostPerMtok, maxCost) + w.lat*norm(c.LatencyMs, maxLat) + w.conc*concS
		if contains(w.tiers, c.Tier) {
			s += tierBoost // the operator's quality judgment
		}
		if c.RateHeadroom != nil {
			s *= 0.5 + 0.5**c.RateHeadroom // back off lanes near their rate cap
		}
		return s
	}
	cands := append([]candidate(nil), p.Candidates...)
	sort.SliceStable(cands, func(i, j int) bool {
		si, sj := score(cands[i]), score(cands[j])
		if si != sj {
			return si > sj // best first
		}
		return cands[i].Idx < cands[j].Idx // deterministic tie-break
	})
	order := make([]int, len(cands))
	for i, c := range cands {
		order[i] = c.Idx
	}
	log.Printf("[smart-router] bucket-weights=%.2f/%.2f/%.2f order=%v (chars=%d tools=%v stream=%v)",
		w.cost, w.lat, w.conc, order, p.Request.TotalChars, p.Request.HasTools, p.Request.Stream)
	return reply{Order: order}
}

func handler(w http.ResponseWriter, r *http.Request) {
	var p payload
	body := reply{Abstain: true} // never 500 on bad input: abstain is the clean path
	if err := json.NewDecoder(r.Body).Decode(&p); err == nil {
		body = rank(p)
	}
	w.Header().Set("Content-Type", "application/json")
	_ = json.NewEncoder(w).Encode(body)
}

func main() {
	addr := "127.0.0.1:8787"
	if len(os.Args) > 1 {
		addr = os.Args[1]
	}
	log.Printf("[smart-router] listening on http://%s/", addr)
	http.HandleFunc("/", handler)
	log.Fatal(http.ListenAndServe(addr, nil))
}

File: examples/smart-router/policy_server.py
#!/usr/bin/env python3
"""Smart-router policy sidecar for busbar (`route: webhook`).

Busbar POSTs a request projection + candidate list before each request's
failover loop; this sidecar classifies the request into a task bucket and
returns a ranked preference `{"order": [idx, ...]}`.

The payload contains ONLY the fields busbar actually sends (see
src/routing/webhook.rs, WebhookRequest): no prompt text ever leaves busbar,
so classification uses shape signals (size, message count, tools, streaming,
max_tokens), not content.

Fail-safe: if this process is slow, down, or wrong, busbar coerces the
decision to the pool's `on_error` (default: weighted SWRR) after
`policy.timeout_ms` (default 150 ms). A broken sidecar never blocks a request.

Run:  python3 policy_server.py [port]   (default 8787)
"""
import json
import sys
from http.server import BaseHTTPRequestHandler, HTTPServer

# ── Task buckets ─────────────────────────────────────────────────────────────
# Weights per bucket: how much each real signal matters, plus which operator
# tiers to boost. "Quality" here is the operator's judgment encoded as `tier`
# on the pool members; busbar enforces it, the sidecar just reads it.
BUCKETS = {
    #                cost  latency  headroom  boosted tiers
    "quick-answer": (0.30, 0.50, 0.20, ("small", "overflow")),
    "code":         (0.10, 0.20, 0.20, ("large", "primary")),
    "long-form":    (0.20, 0.10, 0.20, ("large", "primary")),
    "bulk":         (0.60, 0.10, 0.30, ("small", "overflow")),
}
TIER_BOOST = 0.5  # added when the candidate's tier is in the bucket's list


def classify(req):
    """Bucket the request from the projection busbar sends. No prompt text is
    available (by design), so this is shape-based, in priority order."""
    total_chars = req.get("total_chars") or 0
    max_tokens = req.get("max_tokens") or 0
    if req.get("has_tools"):
        return "code"            # tool/agent traffic wants the capable tier
    if max_tokens >= 4096 or total_chars > 24_000:
        return "long-form"       # ~4 chars/token: 24k chars is roughly 6k tokens
    if not req.get("stream") and req.get("message_count", 0) <= 1:
        return "bulk"            # single-shot, non-interactive: optimize cost
    return "quick-answer"        # interactive default: optimize latency


def score(cand, bucket, max_cost, max_latency, max_conc):
    """Weighted score from the fields busbar provides. Higher is better.
    Missing signals score neutral (0.5) so a cold lane is not punished."""
    w_cost, w_lat, w_conc, tiers = BUCKETS[bucket]
    cost = cand.get("cost_per_mtok")
    cost_s = 0.5 if cost is None else 1.0 - (cost / max_cost if max_cost else 0.0)
    lat = cand.get("latency_ms")  # EWMA; null until the lane has served a request
    lat_s = 0.5 if lat is None else 1.0 - (lat / max_latency if max_latency else 0.0)
    conc_s = (cand.get("available_concurrency", 0) / max_conc) if max_conc else 0.5
    s = w_cost * cost_s + w_lat * lat_s + w_conc * conc_s
    if cand.get("tier") in tiers:
        s += TIER_BOOST
    # Rate headroom, when governance provides it, scales the whole score down
    # as the key nears its RPM/TPM cap: prefer lanes least likely to 429.
    headroom = cand.get("rate_headroom")
    if headroom is not None:
        s *= 0.5 + 0.5 * headroom
    return s


def rank(payload):
    req = payload.get("request", {})
    cands = payload.get("candidates", [])
    if not cands:
        return {"abstain": True}
    bucket = classify(req)
    max_cost = max((c.get("cost_per_mtok") or 0.0) for c in cands) or 0.0
    max_latency = max((c.get("latency_ms") or 0.0) for c in cands) or 0.0
    max_conc = max(c.get("available_concurrency", 0) for c in cands) or 0
    scored = sorted(
        cands,
        key=lambda c: score(c, bucket, max_cost, max_latency, max_conc),
        reverse=True,
    )
    order = [c["idx"] for c in scored]
    print(f"[smart-router] bucket={bucket} order={order} "
          f"(chars={req.get('total_chars')} max_tokens={req.get('max_tokens')} "
          f"tools={req.get('has_tools')} stream={req.get('stream')})")
    return {"order": order}


class Handler(BaseHTTPRequestHandler):
    def do_POST(self):
        try:
            length = int(self.headers.get("Content-Length", 0))
            payload = json.loads(self.rfile.read(length))
            body = rank(payload)
        except Exception as e:  # never 500 on bad input: abstain is the clean path
            print(f"[smart-router] error: {e}; abstaining", file=sys.stderr)
            body = {"abstain": True}
        data = json.dumps(body).encode()
        self.send_response(200)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(data)))
        self.end_headers()
        self.wfile.write(data)

    def log_message(self, *args):  # keep stdout for the ranking log line only
        pass


if __name__ == "__main__":
    port = int(sys.argv[1]) if len(sys.argv) > 1 else 8787
    print(f"[smart-router] listening on http://127.0.0.1:{port}/")
    HTTPServer(("127.0.0.1", port), Handler).serve_forever()
